src/lib/tuning/tuner.cpp

```diff
@@ -0,0 +1,199 @@
+
+void Tuner::wait_for_completion() {
+  if (_evaluate_task) {
+    _evaluate_task->join();
+  }
+  if (_select_task) {
+    _select_task->join();
+  }
+  if (_execute_task) {
+    _execute_task->join();
+  }
+}
+
+void Tuner::_evaluate() {
+  LOG_INFO("Begin tuning evaluation phase...");
+
+  const auto begin = RuntimeClock::now();
+
+  _choices.clear();
+  for (const auto& evaluator : _evaluators) {
+    evaluator->evaluate(_choices);
+
+    const auto runtime = RuntimeClock::now() - begin;
+    if (runtime > _remaining_time_budget) {
+      LOG_INFO("Interrupt evaluation phase: Time budget exceeded.");
+      _status = Status::Timeout;
+      return;
+    }
+    if (runtime > _evaluate_time_budget) {
+      LOG_INFO("Interrupt evaluation phase: Evaluate time budget exceeded.");
+      _status = Status::EvaluationTimeout;
+      return;
+    }
+  }
+  _remaining_time_budget -= RuntimeClock::now() - begin;
+
+  _log_choices();
+  LOG_INFO("Tuning evaluation phase completed.");
+}
+
+void Tuner::_select() {
+  if (_status != Status::Running) {
+    LOG_INFO("Skip tuning selection phase because previous step exceeded the time budget.");
+    return;
+  }
+  LOG_INFO("Begin tuning selection phase...");
+
+  const auto begin = RuntimeClock::now();
+
+  _operations = _selector->select(_choices, _cost_budget);
+
+  const auto runtime = RuntimeClock::now() - begin;
+  if (runtime > _remaining_time_budget) {
+    LOG_INFO("Interrupt selection phase: Time budget exceeded.");
+    _status = Status::Timeout;
+    return;
+  }
+  if (runtime > _select_time_budget) {
+    LOG_INFO("Interrupt selection phase: Select time budget exceeded.");
+    _status = Status::SelectionTimeout;
+    return;
+  }
+  _remaining_time_budget -= runtime;
+
+  _log_operations();
+  LOG_INFO("Tuning selection phase completed.");
+}
+
+void Tuner::_execute() {
+  if (_status != Status::Running) {
+    LOG_INFO("Skip tuning execution phase because previous step exceeded the time budget.");
+    return;
+  }
+  LOG_INFO("Begin tuning execution phase...");
+
+  const auto begin = RuntimeClock::now();
+
+  for (auto& operation : _operations) {
+    operation->execute();
+
+    const auto runtime = RuntimeClock::now() - begin;
+    if (runtime > _remaining_time_budget) {
+      LOG_INFO("Interrupt execution phase: Time budget exceeded.");
+      _status = Status::Timeout;
+      return;
+    }
+    if (runtime > _execute_time_budget) {
+      LOG_INFO("Interrupt execution phase: Execute time budget exceeded.");
+      _status = Status::ExecutionTimeout;
+      return;
+    }
+  }
+
+  _status = Status::Completed;
+  LOG_INFO("Tuning execution phase completed.");
+}
+
+void Tuner::_log_choices() const {
+  LOG_DEBUG("TuningChoice set:");
+  for (const auto& choice : _choices) {
+    LOG_DEBUG("-> " << *choice);
+    // Silence warning about unused variable in release builds
+    (void)choice;
+  }
+}
+
+void Tuner::_log_operations() const {
```

Comment:
Printing all operations and all choices might have been helpful for you, but if all components printed that many information, none of it would be helpful anymore. If, at some point, someone else needs a list of all the operations/choices, writing the for loop that prints them costs them close to no time. I suggest removing most (if not all) of the info and debug prints. Once we have a proper self-tuning-logging-facility, we (i.e., whoever writes that part) \\can bring them bring in properly.

Reply:
We removed the logging code.